src/librustc_typeck/check/mod.rs

```diff
@@ -1371,7 +1370,24 @@ fn check_fn<'a, 'tcx>(
 
     inherited.tables.borrow_mut().liberated_fn_sigs_mut().insert(fn_id, fn_sig);
 
-    fcx.check_return_expr(&body.value);
+    if let ty::Dynamic(..) = declared_ret_ty.kind {
```

Comment:
Can we move this into `check_return_expr` instead?

Reply:
I didn't wan to affect the other call of `check_return_expr` (IIRC it was for closures), but I can try it out and see if that affects anything.

Reply:
Oh, rereading this we can't really move it into `check_return_expr` because in `return x;` `X: ?Sized`, we only require `Sized` from the actual return type in the signature, so I think if I move the `fcx.require_type_is_sized(declared_ret_ty, decl.output.span(), traits::SizedReturnType);` into `check_return_expr` I'll produce undesirable errors (but I am trying it anyways).

Reply:
I think I would be happy if we did at least a `delay_span_bug` in this branch to be sure that if something goes out of sync, we still have an error.

Reply:
As written we are still performing the same checks in both cases, just in the two possible orders. Where would the `delay_span_bug` go? In the first branch? If so, don't see how that is more of a marker for future contributors than the giant comment ^_^.

Reply:
In the first branch; You're basing the fact that we can do this here on that the first branch will always error, and I'm slightly concerned that the changes in the first branch could suddenly stop erroring over the years, so a `delay_span_bug` would be an insurance policy.

Reply:
done